pkg/handler/collector/git/git.go

```diff
@@ -0,0 +1,179 @@
+	"github.com/guacsec/guac/pkg/handler/collector"
+	"github.com/guacsec/guac/pkg/handler/collector/file"
+	"github.com/guacsec/guac/pkg/handler/processor"
+	"github.com/guacsec/guac/pkg/logging"
+	"go.uber.org/zap"
+)
+
+const (
+	CollectorGit = "GIT"
+)
+
+type gitDocumentCollector struct {
+	url           string
+	dir           string
+	lastChecked   time.Time
+	poll          bool
+	interval      time.Duration
+	fileCollector *file.FileCollector
+}
+
+func NewGitCol(ctx context.Context, url string, dir string, poll bool, interval time.Duration) *gitDocumentCollector {
+	logger := logging.FromContext(ctx)
+	fileCollector := file.NewFileCollector(ctx, dir, false, time.Second)
+	err := collector.RegisterDocumentCollector(fileCollector, file.FileCollectorType)
+	if err != nil {
+		logger.Errorf("unable to register file collector: %v", err)
+	}
+
+	return &gitDocumentCollector{
+		url:           url,
+		dir:           dir,
+		poll:          poll,
+		interval:      interval,
+		fileCollector: fileCollector,
+	}
+}
+
+// RetrieveArtifacts collects the documents from the collector. It emits each collected
+// document through the channel to be collected and processed by the upstream processor.
+// The function should block until all the artifacts are collected and return a nil error
+// or return an error from the collector crashing. This function can keep running and check
+// for new artifacts as they are being uploaded by polling on an interval or run once and
+// grab all the artifacts and end.
+func (g *gitDocumentCollector) RetrieveArtifacts(ctx context.Context, docChannel chan<- *processor.Document) error {
+	logger := logging.FromContext(ctx)
+
+	if g.poll {
+		for {
+			if ctx.Err() != nil {
+				return nil
+			}
+			err := g.createOrPull(ctx, logger, docChannel)
+			if err != nil {
+				return err
+			}
+			g.lastChecked = time.Now()
+			time.Sleep(g.interval)
+		}
+	} else {
+		err := g.createOrPull(ctx, logger, docChannel)
+		if err != nil {
+			return err
+		}
+		g.lastChecked = time.Now()
+	}
+
+	return nil
+}
+
+func (g *gitDocumentCollector) createOrPull(ctx context.Context, logger *zap.SugaredLogger, docChannel chan<- *processor.Document) error {
+	exists := checkIfDirExists(g.dir)
+
+	if !exists {
+		if err := os.Mkdir(g.dir, os.ModePerm); err != nil {
+			return err
+		}
+		err := cloneRepoToDir(logger, g.url, g.dir)
+		if err != nil {
+			return err
+		}
+		err = g.fileCollector.RetrieveArtifacts(ctx, docChannel)
+		if err != nil {
+			return err
+		}
+	} else {
+		err := pullRepo(logger, g.dir)
+		if err != nil && err != git.NoErrAlreadyUpToDate {
+			return err
+		} else if err == nil {
+			err = g.fileCollector.RetrieveArtifacts(ctx, docChannel)
+			if err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+// Type returns the collector type
+func (g *gitDocumentCollector) Type() string {
+	return CollectorGit
+}
+
+func checkIfDirExists(name string) bool {
+	_, err := os.Stat(name)
+	return !os.IsNotExist(err)
```

Comment:
i believe this function behavior isn't defined for err == nil, so we may want to explicitly check for that